Migrate all buttons to the new style

mtthwhggns opened the pull request:
# Description

There are two commits, which should probably stay separate.

The first is a tiny bugfix for buttons, where in some browsers, link styles could leak in and look strange.

The second replaces most instances of `btn` with an instance of `aplus-button-*`. There are a few that were skipped, and they are detailed in [the buttons issue](https://github.com/apluslms/a-plus/issues/601). 

# Testing

This only affects styles, so it's been tested manually. 

~~At least one Selenium test is broken, which is why this is currently a DRAFT.~~ The same tests are broken here as on `master`.

# Have you updated the README or other relevant documentation?

*Don't forget to update the documentation if it is relevant for this PR.*

# Is it [Done](https://wiki.aalto.fi/display/EDIT/Definition+of+Done)?

*Clean up your git commit history before submitting the pull request!*

- [ ] I (developer) have created unit tests and the tests pass
- [ ] I (developer) have created functional tests (Selenium tests) if applicable
- [X] I (developer) have tested the changes manually
- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

raphendyr commented on assets/sass/components/_button.scss:
```diff
@@ -10,22 +10,40 @@ $large-height: 24px;
+			$background-color,
+			$aplus-foreground,
+			$aplus-background
+		)
```
I think there could be comma here, but on the other hand, I don't think this function will see more arguments

mtthwhggns replied:
I don't have a strong feeling on the style, but I did this now.

Mankro commented on locale/fi/LC_MESSAGES/django.po:
```diff
@@ -2523,7 +2523,7 @@ msgstr "Lataa"
 "\n"
 "    %(limit)s latest submissions.\n"
-"    <a class=\"btn btn-default btn-sm\" href=\"%(url)s\">Show all</a>\n"
+"    <a class=\"aplus-button--secondary aplus-button--sm\" href=\"%(url)s\">Show all</a>\n"
```
You must replicate this change in the Finnish translation below.

mtthwhggns replied:
Thanks, missed that one!

Mankro replied:
It looks like you still didn't commit the change in the django.po file. This is important to fix.

Mankro commented on news/templates/news/list.html:
```diff
@@ -55,7 +55,7 @@
 <br />
 <p>
-	<a class="btn btn-primary btn-xs" href="{{ instance|url:'news-add' }}">
+	<a class="aplus-button--secondary aplus-button--sm" href="{{ instance|url:'news-add' }}">
```
There are a few pages like this one that now have no primary actions. I was wondering why adding new items can't be the primary action. Is it so because the users may often come to this page for editing something and adding therefore isn't always the most used default action?

mtthwhggns replied:
I never meant to indicate that adding a new item can't be the default action, just that I didn't know that it should be. Pages can have "at most one" default action, so zero is ok too when there are several equal options (login methods for example). 

I don't know enough about how this page is used, but if you believe that most users landing on this page will want to create a new one, I'd be happy to swap it. If most will want to do something else (look at the list, edit or delete one), then maybe it makes more sense to not have a default action on this page. 

What do you think in this case?

Mankro replied:
I see that drawing the line is sometimes blurry. That's OK. Regarding the news, I have the feeling that most teachers would usually add more news items and usually not modify existing news items.

mtthwhggns replied:
The news page now uses `aplus-button--default` for the "New" button as suggested.

Mankro reviewed (changes requested):
Great work. I have some questions and one clear mistake.

Did you test all pages manually with your changes (there are a lot of pages)?

mtthwhggns commented:
> Did you test all pages manually with your changes (there are a lot of pages)?

Nearly all. It took just over a day. There were a few buttons (like the "early access") where I didn't find documentation on how to make them appear, so copied the code out of the conditional blocks temporarily to see how they look, then copied the updated ones back in. 

Thanks for the feedback and spotting the missed one. I'll work through these later today.

mtthwhggns commented on course/templates/course/_siblings.html:
```diff
@@ -10,14 +10,14 @@
         {{ next.name|parse_localization }} &raquo;
         {% if is_course_staff %}
-        <a href="{{ next.link }}" class="btn btn-default btn-xs">
+        <a href="{{ next.link }}" class="aplus-button--secondary aplus-button--xs">
```
That's a good point. 

Fixed now (although the link text is a little cryptic, in my opinion).

Mankro replied:
The "early access" links are someone's historical decision. I think it would be better to just render the link in the exercise title to teachers without a separate "teacher's early access link". But we won't change that in this PR.

mtthwhggns commented on deviations/templates/deviations/list_dl.html:
```diff
@@ -13,7 +13,7 @@
 <br />
 <p>
-	<a class="btn btn-primary btn-xs" href="{{ instance|url:'deviations-add-dl' }}">
+	<a class="aplus-button--secondary aplus-button--xs" href="{{ instance|url:'deviations-add-dl' }}">
```
As a rule, if it's an `a` and it behaves like a button, it should have `role="button"`.

This task was originally about tweaking the classes, so I didn't plan to do them at all in this PR, but I did the ones I noticed. I'm happy to do as many of them as I spot as I check through these.

Mankro replied:
All right, now there are many more role attributes!

mtthwhggns commented on exercise/templates/exercise/_file_link.html:
```diff
@@ -7,7 +7,7 @@
 {% else %}
 	<span>{{ file.filename }}</span>
-	<span class="btn btn-default btn-xs">
+	<span class="aplus-button--secondary aplus-button--xs">
```
> Are the old btn classes preserved on purpose in the first if branch in this file?

No, I just missed it.

> This "button" does not do anything. Is this OK with the new button guideline? I'm curious about the boundaries. Of course, the button style looks good here.

I swapped this one for a `button` with a `disabled` attribute. Not perfect, but I agree that it looks OK.

raphendyr replied:
Yeah, the `disabled` sounds the right thing for this.

mtthwhggns commented on news/templates/news/list.html:
```diff
@@ -55,7 +55,7 @@
 <br />
 <p>
-	<a class="btn btn-primary btn-xs" href="{{ instance|url:'news-add' }}">
+	<a class="aplus-button--primary aplus-button--sm" href="{{ instance|url:'news-add' }}">
```
```suggestion
	<a class="aplus-button--default aplus-button--sm" href="{{ instance|url:'news-add' }}">
```

Mankro replied:
Yikes! The invalid `aplus-button--primary` is back after force-pyshing.

Mankro commented on exercise/templates/exercise/_user_results.html:
```diff
@@ -146,14 +146,14 @@ <h3 class="panel-title">
         <td>
           {% if not accessible %}
-          <a class="btn btn-default btn-xs" href="{{ entry.link }}">
+          <a class="aplus-button--secondary aplus-button--xs" role="button" href="{{ entry.link }}">
```
Now that you removed button classes from "early access" links, this one was probably an oversight since it still has button classes.

mtthwhggns replied:
Thanks, I fixed it now. It seems to be the ones where there are multiple in one file that I'm liable to miss!

Mankro reviewed (commented):
I think one "early access" link has the button class even though you started to remove those. The change in the django.po seems to be missing (HTML classes in the Finnish string).

Mankro commented:
@mtthwhggns After you have fixed the couple of remaining small oversights, you can rebase and squash. Then it is time to merge!

mtthwhggns commented:
> @mtthwhggns After you have fixed the couple of remaining small oversights, you can rebase and squash. Then it is time to merge!

Thanks.

I thought I did them all before, but I guess I'm touching so many files something has to slip through. I've done them now, rebased and squashed.

Mankro reviewed (approved):
Small fixes done. Time to merge.
